ext/app/electron/LoginSystem.ts

```diff
@@ -0,0 +1,119 @@
+import { GristLoginMiddleware, GristLoginSystem, GristServer, setUserInSession } from "app/server/lib/GristServer";
+import { ApiError } from "app/common/ApiError";
+import { Request } from "express";
+import { UserProfile } from "app/common/LoginSessionAPI";
+import cookie from "cookie";
+import { expressWrap } from "app/server/lib/expressWrap";
+import { getOrgUrl } from "app/server/lib/requestUtils";
+import { makeId } from "app/server/lib/idUtils";
+
+
+export type GristDesktopAuthMode = "strict" | "none" | "mixed";
+
+export function getProfile(): UserProfile {
+  // Both variables are guaranteed to be set when this function is invoked,
+  // since loadConfig() is called before a GristApp instance is created.
+  // If they are not set by the user, default values will be used. See config.ts for details.
+  return {
+    email: process.env.GRIST_DEFAULT_EMAIL as string,
+    name: process.env.GRIST_DEFAULT_USERNAME as string,
+  };
+}
+
+// Login and logout, redirecting immediately back.  Signup is treated as login,
+// no nuance here.
+export class ElectronLoginSystem implements GristLoginSystem {
+
+  private static _instance: ElectronLoginSystem;
+
+  private authMode: GristDesktopAuthMode;
+  private credential: string;
+
+  private constructor() {
+    this.credential = makeId();
+    this.authMode = process.env.GRIST_DESKTOP_AUTH as GristDesktopAuthMode;
+  }
+
+  public authenticateURL(url: URL) {
+    const newUrl = new URL(url);
+    if (this.authMode !== "none") {
+      newUrl.searchParams.set("electron_key", this.credential);
+    }
+    return newUrl;
+  }
+
+  public static get instance() {
+    if (!ElectronLoginSystem._instance) {
+      ElectronLoginSystem._instance = new ElectronLoginSystem();
+    }
+    return ElectronLoginSystem._instance;
+  }
+
```

Comment:
I think we could keep it like it was, and pass a credential in, rather than having this generate it? Makes this class:
- More flexible
- Not need to be a singleton

Reply:
On the contrary, I find it tedious to keep references to non-singleton core application components everywhere. Taking this as an example, LoginSystem is made a singleton because WindowManager needs LoginSystem to authenticate document URLs. I don't want to add a `LoginSystem` reference to `WindowManager` for this purpose. That would cause many new issues: How should we initialize these reference fields? In what order should we construct these components?

I think any class that will only have one instance _should_ be made a singleton class. I haven't done so for DocRegistry and WindowManager, but they really should also be.

Reply:
Singletons cause other issues though, such as with testing or with managing data dependencies.

I think a fair number of maintainability problems Grist has have stemmed from overuse of singletons and objects created at module load time, as there's no easy way to parameterise them.